tracing-subscriber/src/filter/env/field.rs

```diff
@@ -275,6 +286,15 @@ impl SpanMatch {
 }
 
 impl<'a> Visit for MatchVisitor<'a> {
+    fn record_f64(&mut self, field: &Field, value: f64) {
+        match self.inner.fields.get(field) {
+            Some((ValueMatch::F64(ref e), ref matched)) if value == *e => {
```

Comment:
We probably need special behavior for `NaN` values here. Two `NaN`s [will never be equal to each other](https://doc.rust-lang.org/1.54.0/std/primitive.f32.html#). However, a user may want to write a filter configuration that enables all spans where a particular value is a `NaN`. Therefore, we probably want to change this behavior.

I would probably change this by adding a separate `ValueMatch::NaN` type, and change the parser to return that instead if `s.parse::<f64>()` returns a value where `is_nan() == true`. Then, in this method, we would add another match arm like
```rust
            Some((ValueMatch::NaN, ref matched)) if value.is_nan() => {
                 matched.store(true, Release);
            }
```
This way, users can match any `NaN` value by writing filters like `{foo=NaN}`, which is probably the desired behavior.

This also avoids the problem of NaNs in the `Ord` impl breaking the total ordering for `ValueMatch`es. Since no `ValueMatch::F64` variants will be created with `NaN` values, we actually _can_ rely on the ordering of `ValueMatch::F64`s being total. We would probably want to add a (debug?) assertion in that case that the values are not actually NaN, to ensure no NaN values sneak in accidentally.

Reply:
@hawkw I had a question for this particular code location, unrelated to the NaN support. Should this method here perform an equality compare against the matching value, or should it do an epsilon compare?

Clippy warns against doing an equality compare but I'm not sure it feels right to be doing an epsilon compare here.

Thoughts?

Reply:
It's probably best to do an epsilon compare here. In most cases, I think the user will be looking for values that are _logically_ "equal" to the provided decimal number, rather than looking for floating-point values whose bit patterns are exactly equal.

Either way, it's probably worth documenting how floats are compared in the documentation for `EnvFilter`, whether we do an exact comparison or an epsilon comparison.